### PR TITLE
OADP-4750: configure OADP NodeAgents on nodes with worker, infra or custom labels

### PR DESCRIPTION
**Pull request by anarnold97**
### JIRA

* [OADP-4750](https://issues.redhat.com/browse/OADP-4750)

### Version(s):

* OCP 4.12+

### Link to docs preview:

* [AWS - Configuring node agents and node labels](https://81145--ocpdocs-pr.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/installing/installing-oadp-aws.html#oadp-configuring-node-agents_installing-oadp-aws)

* [AWS - Setting Velero CPU and memory resource allocations](https://81145--ocpdocs-pr.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/installing/installing-oadp-aws.html#oadp-setting-resource-limits-and-requests_installing-oadp-aws)
 
* [Azure - Configuring node agents and node labels](https://81145--ocpdocs-pr.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/installing/installing-oadp-azure#oadp-configuring-node-agents_installing-oadp-azure)

* [Azure - Setting Velero CPU and memory resource allocations](https://81145--ocpdocs-pr.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/installing/installing-oadp-azure#oadp-setting-resource-limits-and-requests_installing-oadp-azure)

* [GCP - Configuring node agents and node labels](https://81145--ocpdocs-pr.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/installing/installing-oadp-gcp#oadp-configuring-node-agents_installing-oadp-gcp)

* [GCP - Setting Velero CPU and memory resource allocations](https://81145--ocpdocs-pr.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/installing/installing-oadp-gcp#oadp-setting-resource-limits-and-requests_installing-oadp-gcp)

* [OpenShift Virtualization - Configuring node agents and node labels](https://81145--ocpdocs-pr.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/installing/installing-oadp-kubevirt#oadp-configuring-node-agents_installing-oadp-kubevirt)

* [MCG - Configuring node agents and node labels](https://81145--ocpdocs-pr.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/installing/installing-oadp-mcg#oadp-configuring-node-agents_installing-oadp-mcg)

* [MCG - Setting Velero CPU and memory resource allocations](https://81145--ocpdocs-pr.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/installing/installing-oadp-mcg#oadp-setting-resource-limits-and-requests_installing-oadp-mcg)

* [ODF - Configuring node agents and node labels](https://81145--ocpdocs-pr.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/installing/installing-oadp-ocs#oadp-configuring-node-agents_installing-oadp-ocs)

* [ODF - - Setting Velero CPU and memory resource allocations](https://81145--ocpdocs-pr.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/installing/installing-oadp-ocs#oadp-setting-resource-limits-and-requests_installing-oadp-ocs)

* [APIs used with OADP  - Configuring node agents and node labels](https://81145--ocpdocs-pr.netlify.app/openshift-enterprise/latest/backup_and_restore/application_backup_and_restore/oadp-api.html#oadp-configuring-node-agents_oadp-api)

### QE review:
- [ ] QE has approved this change.

Additional information:

### CODE REVIEW

File: modules/oadp-configuring-node-agents.adoc [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+For example:
+
+[source,terminal]
+----
+configuration:
+  nodeAgent:
+    enable: true
+    podConfig:
+      nodeSelector:
+        'node-role.kubernetes.io/nodeAgent: ""'
```
Comment (PrasadJoshi12) [inline modules/oadp-configuring-node-agents.adoc:35]:
```suggestion
        node-role.kubernetes.io/nodeAgent: ""
```

File: modules/oadp-configuring-node-agents.adoc [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+The following example is an anti-pattern of `nodeSelector`, and does not work unless both labels, `'node-role.kubernetes.io/infra: ""'` and `'node-role.kubernetes.io/worker: ""'`, are on the node:
+
+[source,terminal]
+----
+    configuration:
+      nodeAgent:
+        enable: true
+        podConfig:
+          nodeSelector:
+            'node-role.kubernetes.io/infra: ""'
```
Comment (PrasadJoshi12) [inline modules/oadp-configuring-node-agents.adoc:47]: please remove single quotes 
```suggestion
            node-role.kubernetes.io/infra: ""
```

File: modules/oadp-configuring-node-agents.adoc [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+The DPA of {oadp-short} uses `nodeSelector` to select which nodes can run the node agent. `nodeSelector` is the simplest recommended form of node selection constraint.
+
+Any label specified must match the labels on each node.
+
+The correct way to run the node agent on any node you choose is for you to label the nodes with a custom label:
+
+[source,terminal]
+----
+'node-role.kubernetes.io/nodeAgent: ""'
```
Comment (PrasadJoshi12) [inline modules/oadp-configuring-node-agents.adoc:23]: Maybe we can add a label command here for more clarification. 

$ oc label node<node-name> node-role.kubernetes.io/nodeAgent=""

File: modules/oadp-configuring-node-agents.adoc [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+[source,terminal]
+----
+    configuration:
+      nodeAgent:
+        enable: true
+        podConfig:
+          nodeSelector:
+            node-role.kubernetes.io/infra: ""
+            'node-role.kubernetes.io/worker: ""'
```
Comment (PrasadJoshi12) [inline modules/oadp-configuring-node-agents.adoc:48]:
```suggestion
            node-role.kubernetes.io/worker: ""
```

File: modules/oadp-configuring-node-agents.adoc (hunk 3 above)
Comment (anarnold97) [inline modules/oadp-configuring-node-agents.adoc:23]:
```suggestion
$ oc label node node-role.kubernetes.io/nodeAgent=""
```

File: modules/oadp-configuring-node-agents.adoc [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+The DPA of {oadp-short} uses `nodeSelector` to select which nodes can run the node agent. `nodeSelector` is the simplest recommended form of node selection constraint.
+
+Any label specified must match the labels on each node.
+
+The correct way to run the node agent on any node you choose is for you to label the nodes with a custom label:
+
+[source,terminal]
+----
+$ oc label node node-role.kubernetes.io/nodeAgent=""
```
Comment (PrasadJoshi12) [inline modules/oadp-configuring-node-agents.adoc:23]: Users will have to provide the name of the node
```suggestion
$ oc label node/<node-name> node-role.kubernetes.io/nodeAgent=""
```

**Review by PrasadJoshi12 — APPROVED** [review]

File: modules/oadp-configuring-node-agents.adoc [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+Any label specified must match the labels on each node.
+
+The correct way to run the node agent on any node you choose is for you to label the nodes with a custom label:
+
+[source,terminal]
+----
+$ oc label node/<node-name> node-role.kubernetes.io/nodeAgent=""
+----
+
+For example:
```
Comment (shubham-pampattiwar) [inline modules/oadp-configuring-node-agents.adoc:26]: Now use the same custom label in `DPA.spec.configuration.nodeAgent.podCOnfig.nodeSelector` that was used for labelling nodes

**Comment by shubham-pampattiwar** [comment]
Shouldn't we add a note for velero deployment nodeSelector as well ?

File: modules/oadp-configuring-node-agents.adoc (hunk 6 above)
Comment (anarnold97) [inline modules/oadp-configuring-node-agents.adoc:26]:
```suggestion
Use the same custom label in the `DPA.spec.configuration.nodeAgent.podConfig.nodeSelector`, which you used for labelling nodes. For example:
```

**Comment by anarnold97** [comment]
> Shouldn't we add a note for velero deployment nodeSelector as well ?

@shubham-pampattiwar - i have included the module underneath the OADP `PodConfig - nodeSelector` property with a link to the module in the `nodeSelector` property description  

See [commit](9352807de0d02f1090d256011abbe08804e7322b)

**Comment by shubham-pampattiwar** [comment]
I mean we have the `nodeSelector` functionality for both Velero deployment and NodeAgent Daemonset. We should document the `PodConfig - nodeSelector` functionality for both of them.

**Comment by anarnold97** [comment]
@shubham-pampattiwar && @PrasadJoshi12 - have updated. added a snippet and a link to the included module. thanks

**Review by shubham-pampattiwar — APPROVED** [review]

**Review by lpettyjo — COMMENTED** [review]
Please squash your commits into one. For info, see https://docs.google.com/document/d/1fLMpK4bqthtFlCwA36yeo2cI-pExFpX76VMnSi_In6Q/edit#heading=h.epes4xf06pzk. Let me know if you need any help.

Otherwise, LGTM!

**Comment by anarnold97** [comment]
@lpettyjo - thanks. squashed

File: modules/oadp-configuring-node-agents.adoc [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+Any label specified must match the labels on each node.
+
+The correct way to run the node agent on any node you choose is for you to label the nodes with a custom label:
+
+[source,terminal]
+----
+$ oc label node/<node-name> node-role.kubernetes.io/nodeAgent=""
+----
+
+Use the same custom label in the `DPA.spec.configuration.nodeAgent.podConfig.nodeSelector`, which you used for labelling nodes. For example:
```
Comment (adellape) [inline modules/oadp-configuring-node-agents.adoc:27]: We have way more instances of "labeling" across the repo vs "labelling", so suggest switching this one:
```suggestion
Use the same custom label in the `DPA.spec.configuration.nodeAgent.podConfig.nodeSelector`, which you used for labeling nodes. For example:
```

File: modules/oadp-configuring-node-agents.adoc [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+----
+configuration:
+  nodeAgent:
+    enable: true
+    podConfig:
+      nodeSelector:
+        node-role.kubernetes.io/nodeAgent: ""
+----
+
+The following example is an anti-pattern of `nodeSelector`, and does not work unless both labels, `'node-role.kubernetes.io/infra: ""'` and `'node-role.kubernetes.io/worker: ""'`, are on the node:
```
Comment (adellape) [inline modules/oadp-configuring-node-agents.adoc:39]:
```suggestion
The following example is an anti-pattern of `nodeSelector` and does not work unless both labels, `'node-role.kubernetes.io/infra: ""'` and `'node-role.kubernetes.io/worker: ""'`, are on the node:
```

File: snippets/oadp-nodeselector-snippet.adoc [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+//
+// * backup_and_restore/application_backup_and_restore/installing/installing-oadp-aws.adoc
+// * backup_and_restore/application_backup_and_restore/installing/installing-oadp-azure.adoc
+// * backup_and_restore/application_backup_and_restore/installing/installing-oadp-gcp.adoc
+// * backup_and_restore/application_backup_and_restore/installing/installing-oadp-mcg.adoc
+// * backup_and_restore/application_backup_and_restore/installing/installing-oadp-ocs.adoc
+
+:_mod-docs-content-type: SNIPPET
+
+Use `nodeSelector` to select which nodes can run the node agent. `nodeSelector` is the simplest recommended form of node selection constraint. Any label specified must match the labels on each node.
```
Comment (adellape) [inline snippets/oadp-nodeselector-snippet.adoc:11]: Similar here for the snippet, or whatever term you like / is most accurate:
```suggestion
Use the `nodeSelector` field to select which nodes can run the node agent. The `nodeSelector` field is the simplest recommended form of node selection constraint. Any label specified must match the labels on each node.
```

**Review by adellape — COMMENTED** [review]
Holding merge for the `xref` comment. Please ping me back when ready for re-merge-review.

File: backup_and_restore/application_backup_and_restore/installing/installing-oadp-aws.adoc [hunk 10]
```diff
@@ -32,13 +32,19 @@ include::modules/oadp-secrets-for-different-credentials.adoc[leveloffset=+2]
 You can configure the Data Protection Application by setting Velero resource allocations or enabling self-signed CA certificates.
 
 include::modules/oadp-setting-resource-limits-and-requests.adoc[leveloffset=+2]
+
+include::snippets/oadp-nodeselector-snippet.adoc[]
+
+For more details, see xref:oadp-configuring-node-agents_installing-oadp-aws[Configuring node agents and node labels].
```
[the remark this answers is not in the document]
Comment (anarnold97) [reply backup_and_restore/application_backup_and_restore/installing/installing-oadp-aws.adoc]: @adellape - Sorry swapping between repos

**Review by adellape — APPROVED** [review]

**Comment by anarnold97** [comment]
@adellape - crap man, so sorry got side-tracked by another ticket!

**Comment by anarnold97** [comment]
@adellape - thanks. will cherry pick for 4.12 later

**Pull request merged**